src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2Serializer.cs

```diff
@@ -2250,6 +2466,172 @@ protected virtual void WriteSubjectLocality(XmlWriter writer, Saml2SubjectLocali
+            }
+
+            var cryptoProviderFactory = encryptingCredentials.CryptoProviderFactory ?? encryptingCredentials.Key.CryptoProviderFactory;
+
+            if (cryptoProviderFactory == null)
+                throw LogExceptionMessage(new ArgumentException(LogMessages.IDX13600));
+
+            if (!cryptoProviderFactory.IsSupportedAlgorithm(algorithm, sessionKey))
+                throw LogExceptionMessage(new Saml2SecurityTokenEncryptedAssertionEncryptionException(FormatInvariant(LogMessages.IDX13601, algorithm, sessionKey)));
+
+            // Encrypt assertion data
+            AuthenticatedEncryptionResult authenticatedEncryptionResult = null;
+            AuthenticatedEncryptionProvider authenticatedEncryptionProvider = null;
+            try
+            {
+                authenticatedEncryptionProvider = cryptoProviderFactory.CreateAuthenticatedEncryptionProvider(sessionKey, algorithm);
+
+                if (authenticatedEncryptionProvider == null)
+                    throw new Saml2SecurityTokenEncryptedAssertionEncryptionException();
+            }
+            catch (Exception ex)
+            {
+                throw LogExceptionMessage(new Saml2SecurityTokenEncryptedAssertionEncryptionException(LogMessages.IDX13602, ex));
+            }
+
+            try
+            {
+                authenticatedEncryptionResult = authenticatedEncryptionProvider.Encrypt(assertionData, null);
+            }
+            catch (Exception ex)
+            {
+                throw LogExceptionMessage(new Saml2SecurityTokenEncryptedAssertionEncryptionException(FormatInvariant(LogMessages.IDX13603, algorithm, sessionKey), ex));
+            }
+
+            // Populate EncryptedData
+            encryptedData.CipherData.CipherValue = Utility.ConcatByteArrays(authenticatedEncryptionResult.IV, authenticatedEncryptionResult.Ciphertext, authenticatedEncryptionResult.AuthenticationTag);
+            encryptedData.EncryptionMethod = new EncryptionMethod(algorithm);
+            encryptedData.Type = XmlEncryptionConstants.EncryptedDataTypes.Element;
+            encryptedData.Id = new Saml2Id().Value;
+
+            return encryptedData;
+        }
+
+        private EncryptedKey CreateEncryptedKey(EncryptingCredentials encryptingCredentials, SymmetricSecurityKey sessionKey, string encryptedDataId)
+        {
+            if (encryptingCredentials == null)
+                throw LogExceptionMessage(new Saml2SecurityTokenEncryptedAssertionEncryptionException(LogMessages.IDX13624));
+
+            if (sessionKey == null)
+                throw LogExceptionMessage(new Saml2SecurityTokenEncryptedAssertionEncryptionException(LogMessages.IDX13629));
+
+            // AsymmetricSecurityKey is provided:
+            // New session key will be created to encrypt an assertion
+            // Session key will be wrapped with provided AsymmetricSecurityKey
+            if (encryptingCredentials.Key is AsymmetricSecurityKey)
```

Comment:
>if [](start = 12, length = 2)

consider inverting all your if's moving all the short functions on top (e.g. Symmetric does nothing, and not assymetric does nothing. this will give an easy overview of the various return statements from the function.

Reply:
addressed with: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1025/commits/a562057cfa787c2e5212cd69a9b0deb22be013aa